staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go

```diff
@@ -0,0 +1,163 @@
+type Level string
+
+// Valid audit levels
+const (
+	// LevelNone disables auditing
+	LevelNone Level = "None"
+	// LevelMetadata provides the basic level of auditing.
+	LevelMetadata Level = "Metadata"
+	// LevelRequest provides Metadata level of auditing, and additionally
+	// logs the request object (does not apply for non-resource requests).
+	LevelRequest Level = "Request"
+	// LevelResponse provides Request level of auditing, and additionally
+	// logs the response object (does not apply for non-resource requests).
+	LevelResponse Level = "Response"
+)
+
+// Event captures all the information that can be included in an API audit log.
+type Event struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// AuditLevel at which event was generated
+	Level Level `json:"level" protobuf:"bytes,2,opt,name=level,casttype=Level"`
+
+	// Time the request reached the apiserver.
+	Timestamp metav1.Time `json:"timestamp" protobuf:"bytes,3,opt,name=timestamp"`
+	// Unique audit ID, generated for each request.
+	AuditID types.UID `json:"auditID,omitempty" protobuf:"bytes,4,opt,name=auditID,casttype=k8s.io/apimachinery/pkg/types.UID"`
+	// RequestURI is the request URI as sent by the client to a server.
+	RequestURI string `json:"requestURI" protobuf:"bytes,5,opt,name=requestURI"`
+	// Verb is the kubernetes verb associated with the request.
+	// For non-resource requests, this is identical to HttpMethod.
+	Verb string `json:"verb" protobuf:"bytes,6,opt,name=verb"`
+	// Authenticated user information.
+	User authnv1.UserInfo `json:"user" protobuf:"bytes,7,opt,name=user"`
+	// Impersonated user information.
+	// +optional
+	Impersonate *authnv1.UserInfo `json:"impersonate,omitempty" protobuf:"bytes,8,opt,name=impersonate"`
+	// Source IP, from where the request originates.
+	// +optional
+	SourceIP string `json:"sourceIP,omitempty" protobuf:"bytes,9,opt,name=sourceIP"`
+	// Object reference this request is targeted at.
+	// Does not apply for List-type requests, or non-resource requests.
+	// +optional
+	ObjectRef *apiv1.ObjectReference `json:"objectRef,omitempty" protobuf:"bytes,10,opt,name=objectRef"`
+	// The response status, populated even when the ResponseObject is not a Status type.
+	// For successful responses, this will only include the Code and StatusSuccess.
+	// For non-status type error responses, this will be auto-populated with the error Message.
+	// +optional
+	ResponseStatus *metav1.Status `json:"responseStatus,omitempty" protobuf:"bytes,11,opt,name=responseStatus"`
+
+	// API object from the request, in JSON format. The RequestObject is recorded as-is in the request
+	// (possibly re-encoded as JSON), prior to version conversion, defaulting, admission or
+	// merging. It is an external versioned object type, and may not be a valid object on its own.
+	// Omitted for non-resource requests.  Only logged at RequestObject Level and higher.
+	// +optional
+	RequestBody string `json:"requestBody,omitempty" protobuf:"bytes,12,opt,name=requestBody"`
```

Comment:
It shouldn't be runtime.RawExtension. How will RawExtension be encoded in the protobuf case? As protobuf? Then we are lost in the webhook because we cannot decode it and want probably store json anyway.